### PR TITLE
cdc: add CDC delegate

### PR DESCRIPTION
**Pull request by overvenus**
###  What have you changed?

Add CDC delegate it convert raft commands into change data event.

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No.

###  Does this PR affect `tidb-ansible`?

No.

### CODE REVIEW

**Comment by MyonKeminta** [comment]
CI failed

File: Makefile [hunk 1]
```diff
@@ -264,16 +264,32 @@ format: pre-format
...
+ALLOWED_CLIPPY_LINTS=-A clippy::module_inception -A clippy::needless_pass_by_value -A clippy::cognitive_complexity \
+	-A clippy::unreadable_literal -A clippy::should_implement_trait -A clippy::verbose_bit_mask \
+	-A clippy::implicit_hasher -A clippy::large_enum_variant -A clippy::new_without_default \
+	-A clippy::neg_cmp_op_on_partial_ord -A clippy::too_many_arguments \
+	-A clippy::excessive_precision -A clippy::collapsible_if -A clippy::blacklisted_name \
+	-A clippy::needless_range_loop -A clippy::redundant_closure \
+	-A clippy::match_wild_err_arm -A clippy::blacklisted_name -A clippy::redundant_closure_call \
+	-A clippy::identity_conversion -A clippy::new_ret_no_self
+
+# PROST feature works differently in test cdc and backup package, they need to be checked under their folders.
```
Comment (BusyJay) [inline Makefile]: What's the difference?

File: src/storage/mvcc/reader/scanner/forward.rs [hunk 2]
```diff
@@ -689,19 +689,18 @@ where
     }
 }
 
-#[cfg(test)]
-mod test_util {
+pub mod test_util {
```
Comment (BusyJay) [inline src/storage/mvcc/reader/scanner/forward.rs]: What about adding a new feature configuration for this?

File: components/cdc/src/delegate.rs [hunk 3]
```diff
@@ -4,10 +4,22 @@ use std::mem;
 use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
 use std::sync::Arc;
 
-use futures::sync::mpsc::*;
+#[cfg(not(feature = "prost-codec"))]
 use kvproto::cdcpb::*;
+#[cfg(feature = "prost-codec")]
+use kvproto::cdcpb::{
+    event::{
+        row::OpType as EventRowOpType, Entries as EventEntries, Error as EventError,
```
Comment (BusyJay) [inline components/cdc/src/delegate.rs]: Why have the difference? Can it be solved by protobuf-build?

File: Makefile (hunk 1 above)
Comment (overvenus) [reply Makefile]: tests, cdc and backup are not included in tikv's "prost-codec", so `cargo clippy --feature "prost-codec"` does not work. We have to cd their folder and then run the command.

File: components/cdc/src/delegate.rs (hunk 3 above)
Comment (overvenus) [reply components/cdc/src/delegate.rs]: We could write these alias in kvproto, but I want to merge this PR quickly as it blocks other CDC PRs.

File: src/storage/mvcc/reader/scanner/forward.rs (hunk 2 above)
Comment (overvenus) [reply src/storage/mvcc/reader/scanner/forward.rs]: I really no want to add a new feature because of one single mod. I think TiKV's features are too complex 😕

File: Makefile (hunk 1 above)
Comment (BusyJay) [reply Makefile]: Can they be solved in other way? It certainly contributes to compile time to recompile several times.

File: src/storage/mvcc/reader/scanner/forward.rs (hunk 2 above)
Comment (BusyJay) [reply src/storage/mvcc/reader/scanner/forward.rs]: I think it's a common requests to expose test utils for other module. The features may be reused in the future. Besides it's also beneficial for release build.

**Comment by overvenus** [comment]
PTAL, thanks!

File: components/cdc/src/delegate.rs [hunk 4]
```diff
@@ -64,13 +76,696 @@ impl Downstream {
...
+                    let key = Key::from_encoded(put.take_key()).truncate_ts().unwrap();
+                    let row = rows.entry(key.to_raw().unwrap()).or_default();
+                    decode_default(put.take_value(), row);
+                }
+                other => {
+                    panic!("invalid cf {}", other);
+                }
+            }
+        }
+        let mut entires = Vec::with_capacity(rows.len());
```
Comment (MyonKeminta) [inline components/cdc/src/delegate.rs]: s/entires/entries

File: components/cdc/src/endpoint.rs [hunk 5]
```diff
@@ -48,7 +48,6 @@ pub enum Task {
         downstream_id: DownstreamID,
         entries: EntryBatch,
     },
-    #[cfg(not(validate))]
```
Comment (MyonKeminta) [inline components/cdc/src/endpoint.rs]: Why is this removed?

Comment (overvenus) [reply components/cdc/src/endpoint.rs]: It's actually wrong, it should be `#[cfg(not(feature = "validate")]`, so I think it's better to be removed to reduce mistakes and complexity.

File: src/storage/mvcc/reader/scanner/mod.rs [hunk 6]
```diff
@@ -17,7 +17,7 @@ use crate::storage::kv::{
 use crate::storage::mvcc::{default_not_found_error, Result};
 use crate::storage::txn::{Result as TxnResult, Scanner as StoreScanner};
 
-pub use self::forward::{DeltaScanner, EntryScanner};
+pub use self::forward::{test_util, DeltaScanner, EntryScanner};
```
[the remark this answers is not in the document]
Comment (overvenus) [reply src/storage/mvcc/reader/scanner/mod.rs]: No, I tried but it does not work.
